File: ios/PacketTunnel/SingBoxRunner.swift
import Foundation
import NetworkExtension

final class SingBoxRunner: EmbeddedPacketTunnelEngine {
  let name = "sing-box"
  private var sessionId: String?

  func start(provider: NEPacketTunnelProvider, config: PacketTunnelConnectConfig) throws {
    let configJSON = try SingBoxConfigBuilder.build(from: config)
    sessionId = try EmbeddedCoreBridge.startSingBox(configJSON: configJSON, provider: provider)
  }

  func stop() {
    guard let sessionId else { return }
    EmbeddedCoreBridge.stopSingBox(sessionId: sessionId)
    self.sessionId = nil
  }
}

File: android/app/src/main/kotlin/com/example/inet_app/vpn/SingBoxRunner.kt
package com.example.inet_app.vpn

import android.net.VpnService

class SingBoxRunner : EmbeddedVpnEngine {
    override val name: String = "sing-box"
    private var sessionId: String? = null

    override fun start(service: VpnService, tunFd: Int, config: VpnConnectConfig) {
        val configJson = SingBoxConfigFactory.build(config)
        sessionId = EmbeddedCoreBridge.startSingBox(
            context = service.applicationContext,
            configJson = configJson,
            tunFd = tunFd,
            protectSocket = { socketFd -> service.protect(socketFd) },
        )
    }

    override fun stop() {
        val current = sessionId ?: return
        EmbeddedCoreBridge.stopSingBox(current)
        sessionId = null
    }
}
